### PR TITLE
feat!: implement as pins api, add server, convert to monorepo

### PR DESCRIPTION
**Pull request by achingbrain**
Refactors the `@helia/remote-pinning` module to implement the Pins API from Helia, so we can treat remote pinning transparently.

```TypeScript
import { createHelia } from 'helia'
import { heliaWithRemotePins } from '@helia/remote-pinning'
import drain from 'it-drain'

const helia = heliaWithRemotePins(await createHelia(), {
  endpointUrl: 'https://example.com', // the URI for your pinning provider
  accessToken: 'shh-secret' // the secret token/key given to you by your pinning provider
})

await drain(helia.pins.add(cid, { signal: AbortSignal.timeout(5000) }))

// cid is now pinned remotely
```

Adds a server implementation along with interop tests that assert the client works, and that the server passes the compliance tests.

BREAKING CHANGE: the `@helia/remote-pinning` module returns an instance of Pins augmented with extra arguments

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,97 +1,44 @@
-# @helia/remote-pinning <!-- omit in toc -->
+<p align="center">
+  <a href="https://github.com/ipfs/helia" title="Helia">
+    <img src="https://raw.githubusercontent.com/ipfs/helia/main/assets/helia.png" alt="Helia logo" width="300" />
+  </a>
+</p>
+
+# helia-delegated-routing-v1-http-api
```
Comment (SgtPooki) [inline README.md:7]: I assume this was generated with aegir? why did this line show up

```suggestion
```

File: README.md [hunk 2]
```diff
@@ -1,97 +1,44 @@
...
-const cid = await heliaFs.addBytes(encoder.encode('hello world'))
-const addPinResult = await remotePinner.addPin({
-  cid,
-  name: 'helloWorld'
-})
-```
+## About
 
-### Replace a pin
+This repo contains a server implementation of the IPFS [Delegated Routing V1 HTTP API](https://specs.ipfs.tech/routing/http-routing-v1/) along with a client that can be used to interact with any compliant server implementation.
```
Comment (SgtPooki) [inline README.md:21]: this is incorrect

File: README.md [hunk 3]
```diff
@@ -1,97 +1,44 @@
...
-```
+# Packages
 
-## Lead Maintainer
+- [`packages/client`](https://github.com/ipfs/helia-remote-pinning/tree/main/packages/client) Add remote pinning capabilities to Helia
+- [`packages/interop`](https://github.com/ipfs/helia-remote-pinning/tree/main/packages/interop) Interop tests for the IPFS Pinning Service API client and server powered by Helia
+- [`packages/server`](https://github.com/ipfs/helia-remote-pinning/tree/main/packages/server) An IPFS Pinning Service API server powered by Helia
 
-[SgtPooki](https://github.com/sgtpooki)
+> The Delegated Routing V1 HTTP API powered by Helia
```
Comment (SgtPooki) [inline README.md:28]: not sure where the delegated routing stuff is coming from

```suggestion
> The Delegated Routing V1 HTTP API powered by Helia
```

File: packages/client/src/index.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+/**
+ * Patches the passed Helia node with the remote pins instance, this function
+ * makes dealing with the types a little easier
+ */
+export function heliaWithRemotePins <T extends Libp2p = Libp2p> (helia: HeliaLibp2p<T>, init: HeliaRemotePinnerInit = {}): HeliaWithRemotePins<T> {
+  helia.pins = createRemotePins(helia, init)
+
+  return helia as any
```
Comment (SgtPooki) [inline packages/client/src/index.ts:192]: Can we do 

```suggestion
  return helia satisfies HeliaWithRemotePins<T>
```

File: packages/server/src/routes/index.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
+/**
+ * @packageDocumentation
+ *
+ * Configure your existing Fastify instance with routes that conform to the
+ * [Routing V1 HTTP API](https://specs.ipfs.tech/routing/http-routing-v1/) spec.
```
Comment (SgtPooki) [inline packages/server/src/routes/index.ts:5]: change routing v1 to pinning spec?

File: packages/interop/test/compliance.spec.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    // eslint-disable-next-line no-console
+    console.info('npx', '@ipfs-shipyard/pinning-service-compliance', '-s', `http://localhost:${addresses[0]?.port}`, authtoken)
+    const tests = execa('npx', ['@ipfs-shipyard/pinning-service-compliance', '-s', `http://localhost:${addresses[0]?.port}`, authtoken])
+    tests.stdout?.on('data', buf => {
+      process.stdout.write(buf)
+    })
+    tests.stderr?.on('data', buf => {
+      process.stderr.write(buf)
+    })
+    await tests
```
Comment (SgtPooki) [inline packages/interop/test/compliance.spec.ts:54]: I can't remember off the top of my head if the pinning-service-compliance executable will fail on failure, or just generate a report with failures. I think currently we need to check `tests.stderr` for any `FAIL`s

File: packages/server/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+ * })
+ * await server.register(cors, {
+ *   origin: '*',
+ *   methods: ['GET', 'OPTIONS'],
+ *   strictPreflight: false
+ * })
+ *
+ * const helia = await createHelia()
+ *
+ * // configure Routing V1 HTTP API routes
```
Comment (SgtPooki) [inline packages/server/src/index.ts:47]: routing v1 names should be changed to "pinning api" or similar

**Review by SgtPooki — CHANGES_REQUESTED** [review]
a few "routing v1" verbiage that needs removed / updated to be pinning services API, but otherwise awesome. It's great that we have a helia based pinning api server now

File: README.md (hunk 3 above)
Comment (achingbrain) [reply README.md:28]: I copy/pasted a lot of the repo structure..

File: packages/client/src/index.ts (hunk 4 above)
Comment (achingbrain) [reply packages/client/src/index.ts:192]: Unfortunately not:

```
Type 'HeliaLibp2p<T>' does not satisfy the expected type 'HeliaWithRemotePins<T>'.
  Type 'HeliaLibp2p<T>' is not assignable to type '{ pins: RemotePins; }'.
    The types returned by 'pins.ls(...).next(...)' are incompatible between these types.
      Type 'Promise<IteratorResult<Pin, void>>' is not assignable to type 'Promise<IteratorResult<RemotePin, void>>'.
        Type 'IteratorResult<Pin, void>' is not assignable to type 'IteratorResult<RemotePin, void>'.
          Type 'IteratorYieldResult<Pin>' is not assignable to type 'IteratorResult<RemotePin, void>'.
            Type 'IteratorYieldResult<Pin>' is not assignable to type 'IteratorYieldResult<RemotePin>'.
              Type 'Pin' is not assignable to type 'RemotePin'.ts(1360)
```

File: packages/server/src/index.ts (hunk 7 above)
Comment (achingbrain) [reply packages/server/src/index.ts]: Done!

File: packages/server/src/routes/index.ts (hunk 5 above)
Comment (achingbrain) [reply packages/server/src/routes/index.ts]: Done!

File: README.md (hunk 3 above)
Comment (achingbrain) [reply README.md]: Updated

File: README.md (hunk 2 above)
Comment (achingbrain) [reply README.md]: Fixed

File: packages/interop/test/compliance.spec.ts (hunk 6 above)
Comment (achingbrain) [reply packages/interop/test/compliance.spec.ts]: ..it doesn't, it exits cleanly even if tests have failed.  Turns out the test output is on `stdout` too, not `stderr`, and it's all emojis.

I've added a check.

**Pull request merged**